### PR TITLE
PlacementService: improve performance, reduce directory requests, improve consistency of decisions

### PR DESCRIPTION
**Pull request by ReubenBond**
This PR attempts to improve how PlacementService addresses messages in the following ways:
* Reduce the number of outgoing calls to the grain directory when multiple concurrent requests arrive for the same grain.
* Improve the consistency of decisions made by ensuring that all concurrent requests to a given grain are given the same address

The basic architecture is that PlacementService has a collection of associated PlacementWorker instances (arbitrarily selected to be one per core) which are responsible for asynchronous *address message* requests. Messages are consistently assigned to the same PlacementWorker by hashing the TargetGrain id. Each PlacementWorker batches messages for the same grain together so that decisions will be consistent and directory requests will be deduplicated.

### CODE REVIEW

File: src/Orleans.Runtime/Placement/PlacementService.cs [hunk 1]
```diff
@@ -179,5 +154,192 @@ public SiloAddress[] AllActiveSilos
...
+                    activationId = ActivationId.GetDeterministic(target.GrainIdentity);
+                }
+                else
+                {
+                    activationId = ActivationId.NewId();
+                }
+
+                result = ActivationAddress.GetAddress(siloAddress, targetGrain, activationId);
+                _placementService._grainLocator.InvalidateCache(targetGrain);
+                _placementService._grainLocator.CachePlacementDecision(result);
```
Comment (benjaminpetit) [inline src/Orleans.Runtime/Placement/PlacementService.cs:333]: If the grain doesn't end up activated on the target silo, you will have a wrong entry in the cache for a long time. I think we have the same issue if a grain is deactivated after placement, but I feel here we are making things worse.

I believe right now we only invalidate cache entries form the `LocalGrainDirectory`. We should change that in this PR.

File: src/Orleans.Runtime/Placement/PlacementService.cs [hunk 2]
```diff
@@ -16,6 +17,7 @@ namespace Orleans.Runtime.Placement
     /// </summary>
     internal class PlacementService : IPlacementContext
     {
+        private const int PlacementWorkerCount = 16;
```
Comment (benjaminpetit) [inline src/Orleans.Runtime/Placement/PlacementService.cs:20]: Should this be configurable, or at least based on some core count?

File: src/Orleans.Core/GrainDirectory/IGrainLocator.cs [hunk 3]
```diff
@@ -15,5 +15,10 @@ public interface IGrainLocator
         ValueTask<ActivationAddress> Lookup(GrainId grainId);
 
         bool TryLocalLookup(GrainId grainId, out ActivationAddress addresses);
+
+        void CachePlacementDecision(ActivationAddress address);
+
+        void InvalidateCache(GrainId grainId);
+        bool TryCachedLookup(GrainId grainId, out ActivationAddress address);
```
Comment (benjaminpetit) [inline src/Orleans.Core/GrainDirectory/IGrainLocator.cs:22]: `TryLocalLookup` and `TryCachedLookup` sound very similar...

**Comment by benjaminpetit** [comment]
I wonder if we should just have a dedicated very short-lived cache in the placement service instead of exposing cache method in the grain locator (even though I think we should expose a cache invalidation method in the locator)

File: src/Orleans.Runtime/Placement/PlacementService.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.Runtime/Placement/PlacementService.cs:333]: The entry will be invalidated by return from the call made to the remote silo (forwarded to the right silo, hopefully, or rejected back to the caller)

File: src/Orleans.Runtime/Placement/PlacementService.cs (hunk 2 above)
Comment (ReubenBond) [reply src/Orleans.Runtime/Placement/PlacementService.cs:20]: I suppose we could, but I'm not sure we want to expose it as a tuning knob and I'm not sure a higher or lower value is any better than this effectively random value.

File: src/Orleans.Runtime/Placement/PlacementService.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.Runtime/Placement/PlacementService.cs:333]: Addressed in latest

File: src/Orleans.Core/GrainDirectory/IGrainLocator.cs (hunk 3 above)
Comment (ReubenBond) [reply src/Orleans.Core/GrainDirectory/IGrainLocator.cs:22]: If I remove the 'd' and make it `TryCacheLookup` is that clearer (or less clear)? Alternatively, how about `TryCacheOnlyLookup`?
`TryLocalLookup` really only makes sense for the DHT directory, so I think it's the outlier.

**Comment by ReubenBond** [comment]
> I wonder if we should just have a dedicated very short-lived cache in the placement service instead of exposing cache method in the grain locator (even though I think we should expose a cache invalidation method in the locator)

Possibly, but I am wary of multiple caches making cache coherency even more complicated.

File: src/Orleans.Core/GrainDirectory/IGrainLocator.cs (hunk 3 above)
Comment (benjaminpetit) [reply src/Orleans.Core/GrainDirectory/IGrainLocator.cs:22]: I think we could remove `TryLocalLookup` and keep the new `TryCachedLookup`. The `Lookup` method returns a `ValueTask`, so we could do the specific local lookup for the DHT directory in there.

**Pull request merged**